src/huggingface_hub/commit_api.py

```diff
@@ -0,0 +1,424 @@
+        num_threads (`int`, *optional*):
+            The number of concurrent threads to use when uploading. Defaults to 5.
+
+
+    Raises: `RuntimeError` if an upload failed for any reason
+
+    Raises: `ValueError` if the server returns malformed responses
+
+    Raises: `requests.HTTPError` if the LFS batch endpoint returned an HTTP
+        error
+
+    """
+    # Step 1: retrieve upload instructions from the LFS batch endpoint
+    batch_actions, batch_errors = post_lfs_batch_info(
+        upload_infos=[op.upload_info() for op in additions],
+        token=token,
+        repo_id=repo_id,
+        repo_type=repo_type,
+        revision=revision,
+        endpoint=endpoint,
+    )
+    if batch_errors:
+        message = "\n".join(
+            [
+                f'Encountered error for file with OID {err.get("oid")}:'
+                f' `{err.get("error", {}).get("message")}'
+                for err in batch_errors
+            ]
+        )
+        raise ValueError(f"LFS batch endpoint returned errors:\n{message}")
+
+    # Step 2: upload files concurrently according to these instructions
+    oid2addop = {add_op.upload_info().sha256.hex(): add_op for add_op in additions}
+    with ThreadPoolExecutor(max_workers=num_threads) as pool:
+        logger.debug(
+            f"Uploading {len(batch_actions)} LFS files to the Hub using up to"
+            f" {num_threads} threads concurrently"
+        )
+        # Upload the files concurrently, stopping on the first exception
+        futures2operation: Dict[futures.Future, CommitOperationAdd] = {
+            pool.submit(
+                _upload_lfs_object,
+                operation=oid2addop[batch_action["oid"]],
+                lfs_batch_action=batch_action,
+                token=token,
+            ): oid2addop[batch_action["oid"]]
+            for batch_action in batch_actions
+        }
+        completed, pending = futures.wait(
+            futures2operation,
+            return_when=futures.FIRST_EXCEPTION,
+        )
+
+        for pending_future in pending:
+            # Cancel pending futures
+            # Uploads already in progress can't be stopped unfortunately,
+            # as `requests` does not support cancelling / aborting a request
+            pending_future.cancel()
+
+        for future in completed:
+            operation = futures2operation[future]
+            try:
+                future.result()
+            except Exception as exc:
+                # Raise the first exception encountered
+                raise RuntimeError(
+                    f"Error while uploading {operation.path_in_repo} to the Hub"
+                ) from exc
+
+
+def _upload_lfs_object(
+    operation: CommitOperationAdd, lfs_batch_action: dict, token: str
+):
+    """
+    Handles uploading a given object to the Hub with the LFS protocol.
+
+    Defers to `~.utils.lfs.lfs_upload` for the actual upload logic.
```

Comment:
This pattern (`~.utils.lfs.lfs_upload`) understood by doc builder?

Reply:
Not sure, I switched to `[~huggingface_hub.XXX]` as elsewhere